sermant-plugins/sermant-mq-grayscale/sermant-mq-grayscale-plugin/src/main/java/io/sermant/mq/grayscale/utils/SubscriptionDataUtils.java

```diff
@@ -0,0 +1,413 @@
+        if (StringUtils.isEmpty(MqGrayscaleConfigUtils.getGrayGroupTag())) {
+            // base model return without exclude group message
+            if (CONSUME_TYPE_BASE.equals(MqGrayscaleConfigUtils.getConsumeType())) {
+                List<GrayTagItem> items = MqGrayscaleConfigUtils.getGrayTagItemByExcludeGroupTags();
+                if (!items.isEmpty()) {
+                    sb.append(buildBaseConsumerSql92Expression(items));
+                }
+                return sb.toString();
+            }
+
+            // auto model return without exclude group and current consume message gray group message
+            sb.append(buildBaseConsumerSql92Expression(getAutoTypeGrayTagItems(addrTopicGroupKey)));
+        } else {
+            MqGrayscaleConfig mqGrayscaleConfig = MqGrayscaleConfigUtils.getGrayscaleConfigs();
+            Optional<GrayTagItem> grayTagItem
+                    = mqGrayscaleConfig.getGrayTagByGroupTag(MqGrayscaleConfigUtils.getGrayGroupTag());
+            if (grayTagItem.isPresent()) {
+                sb.append(buildGrayConsumerSql92Expression(grayTagItem.get()));
+            } else {
+                LOGGER.warning(String.format(Locale.ENGLISH, "current gray group [%s] had not set grayscale, set it "
+                        + "and restart service to valid.", MqGrayscaleConfigUtils.getGrayGroupTag()));
+            }
+        }
+        return sb.toString();
+    }
+
+    private static List<GrayTagItem> getAutoTypeGrayTagItems(String addrTopicGroupKey) {
+        List<GrayTagItem> excludeItems = MqGrayscaleConfigUtils.getGrayTagItemByExcludeGroupTags();
+        List<GrayTagItem> autoDiscoveryGrayTags = AUTO_CHECK_GRAY_TAGS.get(addrTopicGroupKey);
+        if (autoDiscoveryGrayTags != null && !autoDiscoveryGrayTags.isEmpty()) {
+            excludeItems.addAll(autoDiscoveryGrayTags);
+        }
+        return excludeItems;
+    }
+
+    private static String buildGrayConsumerSql92Expression(GrayTagItem item) {
+        Map<String, List<String>> trafficTagMap = new HashMap<>();
+        for (Map.Entry<String, String> entry : item.getTrafficTag().entrySet()) {
+            buildTrafficTagMap(trafficTagMap, entry);
+        }
+        StringBuilder builder = new StringBuilder();
+        if (trafficTagMap.size() > 1) {
+            builder.append(LEFT_BRACKET);
+        }
+        for (Map.Entry<String, List<String>> envEntry : trafficTagMap.entrySet()) {
+            if (builder.length() > 1) {
+                builder.append(" or ");
+            }
+            builder.append(LEFT_BRACKET)
+                    .append(envEntry.getKey())
+                    .append(" in ")
+                    .append(getStrForSets(new HashSet<>(envEntry.getValue())))
+                    .append(RIGHT_BRACKET);
+        }
+        if (trafficTagMap.size() > 1) {
+            builder.append(RIGHT_BRACKET);
+        }
+        return builder.length() > 0 ? builder.toString() : "";
```

Comment:
Just return builder.toString(). When builder.length() is 0, builder.toString() is an empty string.

Reply:
fixed